src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,309 @@
+    }
+  }
+}
+source_locationt make_function_pointer_restriction_assertion_source_location(
+  source_locationt source_location,
+  const function_pointer_restrictionst::value_type restriction)
+{
+  std::stringstream comment;
+  comment << restriction.first << " must be ";
+  if(restriction.second.size() == 1)
+  {
+    comment << restriction.second.front();
+  }
+  else
+  {
+    comment << "one of [";
+    bool first = true;
+    for(auto const &target : restriction.second)
+    {
+      if(!first)
+      {
+        comment << ", ";
+      }
+      else
+      {
+        first = false;
+      }
+      comment << target;
+    }
+    comment << ']';
+  }
+  source_location.set_comment(string2id(comment.str()));
+  source_location.set_property_class(ID_assertion);
+  // XXX I want to hide these assertions unless they fail - apparently this is
+  ///    not the correct way to do this?
+  source_location.set_hide();
+  return source_location;
+}
+
+template <typename Handler>
+void for_each_function_call(goto_functiont &goto_function, Handler handler)
+{
+  for_each_goto_location_if(
+    goto_function,
+    [](goto_programt::targett target) { return target->is_function_call(); },
+    [&](goto_programt::targett target) {
+      if(target->is_function_call())
+      {
+        handler(target);
+      }
+    });
+}
+} // namespace
+
+void restrict_function_pointers(
+  goto_modelt &goto_model,
+  const function_pointer_restrictionst &restrictions)
+{
+  typecheck_function_pointer_restrictions(goto_model, restrictions);
+
+  // we iterate over all restrictions we might have
+  for(auto &goto_function : goto_model.goto_functions.function_map)
+  {
+    // for each function call, we l
+    for_each_function_call(
+      goto_function.second, [&](const goto_programt::targett location) {
+        // Check if this is calling a function pointer, and if so if it is one
+        // we have a restriction for
+        const auto &original_function_call = location->get_function_call();
+        if(can_cast_expr<dereference_exprt>(original_function_call.function()))
+        {
+          // because we simplify before this stage a dereference can only
+          // dereference a symbol expression
+          auto const &called_function_pointer =
+            to_dereference_expr(original_function_call.function()).pointer();
+          PRECONDITION(can_cast_expr<symbol_exprt>(called_function_pointer));
+          auto const &pointer_symbol = to_symbol_expr(called_function_pointer);
+          auto const restriction_iterator =
+            restrictions.restrictions.find(pointer_symbol.get_identifier());
+          if(restriction_iterator != restrictions.restrictions.end())
+          {
+            auto const &restriction = *restriction_iterator;
+            auto const is_calling_pointer_symbol =
+              to_dereference_expr(original_function_call.function())
+                .pointer() == pointer_symbol;
+            const symbol_exprt &pointer_symbol =
```

Comment:
First, can we rename this: there already is one `pointer_symbol` above? Second, can we move it below, closer to it's use?